docs/code-quality-tasks/module-graph-import-issues-report.md

```diff
@@ -0,0 +1,191 @@
+
+// Bad
+export default new Clock();
+
+// Good
+export const clock = new Clock();
+// Or better: export a factory function
+export function createClock(): Clock { ... }
+```
+
+## 3. Missing or Incorrect Import Grouping
+
+The following files don't follow the import grouping convention (standard library → external → internal):
+
+### `packages/jumble/src/components/CharmRunner.tsx`
+```typescript
+// Current (mixed imports)
+import { html } from "lit";
+import { Recipe } from "@commontools/common-runner";
+import { SpellRunner } from "./SpellRunner.js";
+
+// Should be:
+// Standard library
+import { html } from "lit";
+
+// External
+// (none in this case)
+
+// Internal
+import { Recipe } from "@commontools/common-runner";
+import { SpellRunner } from "./SpellRunner.js";
+```
+
+### `packages/builder/src/utils.ts`
+```typescript
+// Current (internal packages mixed without grouping)
+import { JSONSchema7 } from "json-schema";
+import { isOpaqueRef, OpaqueRef } from "./spell.js";
+import { diffAndUpdate, maybeGetCellLink } from "@commontools/runner";
+
+// Should be:
+// Standard library
+// (none)
+
+// External
+import { JSONSchema7 } from "json-schema";
+
+// Internal
+import { diffAndUpdate, maybeGetCellLink } from "@commontools/runner";
+import { isOpaqueRef, OpaqueRef } from "./spell.js";
+```
+
+### `packages/identity/src/ed25519/index.ts`
+- External imports (`@scure/bip39`) mixed with internal imports without proper grouping
+
+## 4. Heavy Dependencies in Utility Modules
+
+### `packages/builder/src/utils.ts`
+**Issue**: This utility module imports from `@commontools/runner`, making it a heavyweight utility rather than a lightweight one.
+
+**Impact**: 
+- Any code that needs simple utilities from `builder` must also load the entire `runner` package
+- Creates the circular dependency issue mentioned above
+- Violates the principle that utility modules should have minimal dependencies
+
+**Recommendation**: 
+- Move utilities that depend on `runner` to a separate module like `builder-runner-utils.ts`
+- Keep pure utilities without external dependencies in `utils.ts`
+
+## 5. Module-Specific Dependencies (Good Practices Found)
+
+The codebase correctly follows the guideline for module-specific dependencies:
+
+### Good Examples:
+- `packages/llm/deno.json` - Correctly declares `json5` dependency
+- `packages/ui/deno.json` - Correctly declares `@shoelace-style/shoelace` dependency
+- Most packages don't have unnecessary dependencies in their `deno.json`
+
+**This is a positive finding** - the codebase correctly avoids adding package-specific dependencies to the workspace `deno.json`.
+
+## 6. Non-standard JS Usage
+
+**Positive Finding**: The codebase is clean of non-portable JavaScript patterns:
+- No direct use of environment variables in module scope
+- No Vite-specific APIs (`import.meta.hot`, `import.meta.glob`) outside of Vite plugin code
+- Clean separation between build-time and runtime code
+
+## Recommendations
```

Comment:
The problems identified seem accurate, but the recommendations lack context of a package containing user/recipe-only capabilities, which could solve some of the circular dependencies as well